clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -69,15 +79,18 @@ public Metadata(long refreshBackoffMs, long metadataExpireMs) {
         this.version = 0;
         this.cluster = Cluster.empty();
         this.needUpdate = false;
-        this.topics = new HashSet<String>();
+        // KAFKA-3428: using concurrent set since reads to topics can be concurrent with writes.
+        this.topics = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
         this.listeners = new ArrayList<>();
         this.needMetadataForAllTopics = false;
```

Comment:
I would move all of these apart from the first two to the declaration of the fields themselves.